File: producer/producer.py
from datetime import datetime
from kafka import KafkaProducer  # type: ignore
import json
import time
import random
import threading
import sys
import os

# IP del PC dove gira Kafka
KAFKA_SERVER = '192.168.178.166:29092'

if len(sys.argv) < 2:
    print("Uso: python producer.py <NUM_THREADS>")
    sys.exit(1)

NUM_THREADS = int(sys.argv[1])
print(f"Avvio di {NUM_THREADS} thread...")

stop_event = threading.Event()


class SafeCounter:
    def __init__(self):
        self.sent = 0
        self.failed = 0
        self._lock = threading.Lock()

    def inc_sent(self):
        with self._lock:
            self.sent += 1

    def inc_failed(self):
        with self._lock:
            self.failed += 1

    def snapshot(self):
        with self._lock:
            return self.sent, self.failed


counter = SafeCounter()


def create_producer():
    while True:
        try:
            producer = KafkaProducer(
                bootstrap_servers=[KAFKA_SERVER],
                value_serializer=lambda v: json.dumps(v).encode(  # type: ignore
                    'utf-8'),
                linger_ms=100,        # aspetta max 100ms prima di inviare batch
                batch_size=16384,     # 16 KB per batch
                max_request_size=1048576,  # max 1 MB per richiesta
                acks=1                # meno overhead
            )
            print("Producer Kafka connesso.")
            return producer
        except Exception as e:
            print(f"Errore connessione Kafka: {e}. Riprovo in 1s...")
            time.sleep(1)


def send_temperature(thread_id: int):
    producer = create_producer()
    # count = 0

    while not stop_event.is_set():
        temperatura = 20 + random.randint(0, 10)
        data = {
            "temperatura": str(temperatura),
            "timestamp": datetime.now().strftime("%Y-%m-%dT%H:%M:%S.%f"),
            "thread_id": str(thread_id)
        }

        try:
            producer.send("temperature", data) # type: ignore
            counter.inc_sent()
        except Exception:
            counter.inc_failed()
            producer.close() # type: ignore
            producer = create_producer()

        # count += 1
        # if count % 10 == 0:  # stampa ogni 10 messaggi
        #     sent, failed = counter.snapshot()
        #     print(f"[Thread {thread_id}] sent: {sent}, failed: {failed}")

        time.sleep(0.01)

    producer.close() # type: ignore



threads: list[threading.Thread] = []
for i in range(NUM_THREADS):
    t = threading.Thread(target=send_temperature, args=(i+1,), daemon=True)
    t.start()
    threads.append(t)  # type: ignore

print(f"Avviati {NUM_THREADS} thread. PID: {os.getpid()}")

try:
    stop_event.wait(timeout=600)
    stop_event.set()
    print("\nTempo scaduto: fermo tutti i thread...")
    for t in threads:
        t.join()

except KeyboardInterrupt:
    print("\nTerminazione richiesta dall'utente.")
    stop_event.set()
    for t in threads:
        t.join()

# Statistiche
sent, failed = counter.snapshot()
print("\n=== STATISTICHE ===")
print(f"Messaggi inviati: {sent}")
print(f"Messaggi falliti: {failed}")
print("===================")

File: producer/producer.cpp
#include <rdkafka/rdkafka.h>

#include <atomic>
#include <chrono>
#include <csignal>
#include <ctime>
#include <iomanip>
#include <iostream>
#include <mutex>
#include <random>
#include <sstream>
#include <string>
#include <thread>
#include <vector>

static const std::string KAFKA_SERVER = "192.168.178.166:29092";
static const std::string TOPIC = "temperature";

std::atomic<bool> stop_event(false);

/* =======================
   SafeCounter
   ======================= */
class SafeCounter {
private:
    int sent{0};
    int failed{0};
    std::mutex mtx;

public:
    void inc_sent() {
        std::lock_guard<std::mutex> lock(mtx);
        sent++;
    }

    void inc_failed() {
        std::lock_guard<std::mutex> lock(mtx);
        failed++;
    }

    std::pair<int, int> snapshot() {
        std::lock_guard<std::mutex> lock(mtx);
        return {sent, failed};
    }
};

SafeCounter counter;

/* =======================
   Utility timestamp ISO
   ======================= */
std::string now_iso() {
    auto now = std::chrono::system_clock::now();
    auto t = std::chrono::system_clock::to_time_t(now);
    auto us = std::chrono::duration_cast<std::chrono::microseconds>(
                  now.time_since_epoch())
                  .count() %
              1000000;

    std::ostringstream oss;
    oss << std::put_time(std::localtime(&t), "%Y-%m-%dT%H:%M:%S")
        << "." << std::setw(6) << std::setfill('0') << us;
    return oss.str();
}

/* =======================
   Create Kafka producer
   ======================= */
rd_kafka_t* create_producer() {
    char errstr[512];

    while (true) {
        rd_kafka_conf_t* conf = rd_kafka_conf_new();

        if (rd_kafka_conf_set(conf, "bootstrap.servers",
                              KAFKA_SERVER.c_str(),
                              errstr, sizeof(errstr)) != RD_KAFKA_CONF_OK) {
            std::cerr << "Errore config Kafka: " << errstr << std::endl;
            rd_kafka_conf_destroy(conf);
            std::this_thread::sleep_for(std::chrono::seconds(1));
            continue;
        }

        rd_kafka_t* producer = rd_kafka_new(
            RD_KAFKA_PRODUCER, conf, errstr, sizeof(errstr));

        if (!producer) {
            std::cerr << "Errore connessione Kafka: "
                      << errstr << ". Riprovo...\n";
            std::this_thread::sleep_for(std::chrono::seconds(1));
            continue;
        }

        std::cout << "Producer Kafka connesso.\n";
        return producer;
    }
}

/* =======================
   Thread worker
   ======================= */
void send_temperature(int thread_id) {
    rd_kafka_t* producer = create_producer();
    std::mt19937 rng(std::random_device{}());
    std::uniform_int_distribution<int> dist(20, 30);

    while (!stop_event.load()) {
        int temperatura = dist(rng);

        std::ostringstream json;
        json << "{"
             << "\"temperatura\":" << temperatura << ","
             << "\"timestamp\":\"" << now_iso() << "\","
             << "\"thread_id\":" << thread_id
             << "}";

        std::string payload = json.str();

        rd_kafka_resp_err_t err = rd_kafka_producev(
            producer,
            RD_KAFKA_V_TOPIC(TOPIC.c_str()),
            RD_KAFKA_V_MSGFLAGS(RD_KAFKA_MSG_F_COPY),
            RD_KAFKA_V_VALUE(payload.c_str(), payload.size()),
            RD_KAFKA_V_END);

        if (err) {
            counter.inc_failed();
            std::cerr << "[Thread " << thread_id
                      << "] Errore invio: "
                      << rd_kafka_err2str(err)
                      << ". Riconnessione...\n";

            rd_kafka_flush(producer, 5000);
            rd_kafka_destroy(producer);
            producer = create_producer();
        } else {
            rd_kafka_flush(producer, 1000);
            counter.inc_sent();
            std::cout << "[Thread " << thread_id
                      << "] Inviato: " << payload << "\n";
        }

        auto [sent, failed] = counter.snapshot();
        std::cout << "[Thread " << thread_id
                  << "] sent: " << sent
                  << ", failed: " << failed << "\n";

        std::this_thread::sleep_for(std::chrono::seconds(1));
    }

    rd_kafka_flush(producer, 5000);
    rd_kafka_destroy(producer);
    std::cout << "[Thread " << thread_id << "] Terminato.\n";
}

/* =======================
   Signal handler
   ======================= */
void handle_signal(int) {
    stop_event.store(true);
}

/* =======================
   MAIN
   ======================= */
int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Uso: ./producer <NUM_THREADS>\n";
        return 1;
    }

    int NUM_THREADS = std::stoi(argv[1]);
    std::cout << "Avvio di " << NUM_THREADS << " thread...\n";

    signal(SIGINT, handle_signal);

    std::vector<std::thread> threads;
    for (int i = 0; i < NUM_THREADS; ++i) {
        threads.emplace_back(send_temperature, i + 1);
    }

    std::this_thread::sleep_for(std::chrono::seconds(600));
    stop_event.store(true);

    for (auto& t : threads)
        t.join();

    auto [sent, failed] = counter.snapshot();
    std::cout << "\n=== STATISTICHE ===\n";
    std::cout << "Messaggi inviati: " << sent << "\n";
    std::cout << "Messaggi falliti: " << failed << "\n";
    std::cout << "===================\n";

    return 0;
}
